consensus/src/simplex/state.rs

```diff
@@ -0,0 +1,2331 @@
+///   [`Actor::verified`](crate::simplex::actors::voter::Actor::verified).
+///
+/// Keeping these flows centralized in the round state lets tests and recovery logic manipulate
+/// proposals without needing to instantiate the async actor.
+#[derive(Default)]
+struct ProposalSlot<D>
+where
+    D: Digest,
+{
+    proposal: Option<Proposal<D>>,
+    status: ProposalStatus,
+    requested_build: bool,
+    requested_verify: bool,
+    awaiting_parent: Option<View>,
+}
+
+impl<D> ProposalSlot<D>
+where
+    D: Digest + Clone + PartialEq,
+{
+    fn new() -> Self {
+        Self {
+            proposal: None,
+            status: ProposalStatus::None,
+            requested_build: false,
+            requested_verify: false,
+            awaiting_parent: None,
+        }
+    }
+
+    fn proposal(&self) -> Option<&Proposal<D>> {
+        self.proposal.as_ref()
+    }
+
+    fn status(&self) -> ProposalStatus {
+        self.status
+    }
+
+    fn should_build(&self) -> bool {
+        !self.requested_build && self.proposal.is_none()
+    }
+
+    fn set_building(&mut self) {
+        self.requested_build = true;
+        self.awaiting_parent = None;
+    }
+
+    fn request_verify(&mut self) -> bool {
+        if self.requested_verify {
+            return false;
+        }
+        self.requested_verify = true;
+        true
+    }
+
+    /// Marks the slot as waiting on parent certificates.
+    ///
+    /// Returns `true` the first time it is invoked so callers can distinguish
+    /// between a freshly-discovered gap (which should trigger a resolver fetch)
+    /// and repeated checks we expect to run while we wait for the data.
+    fn mark_parent_missing(&mut self, parent: View) -> bool {
+        match self.awaiting_parent {
+            Some(missing) if missing == parent => false,
+            None | Some(_) => {
+                self.awaiting_parent = Some(parent);
+                true
+            }
+        }
+    }
+
+    fn clear_parent_missing(&mut self) {
+        self.awaiting_parent = None;
+    }
+
+    fn record_proposal(&mut self, replay: bool, proposal: Proposal<D>) {
+        if let Some(existing) = &self.proposal {
+            if !replay {
+                debug!(
+                    ?existing,
+                    ?proposal,
+                    "ignoring local proposal because slot already populated"
+                );
+                return;
+            }
+        }
+        self.proposal = Some(proposal);
+        self.status = ProposalStatus::Verified;
+        self.requested_build = true;
+        self.requested_verify = true;
+    }
+
+    fn mark_verified(&mut self) -> bool {
+        if self.status != ProposalStatus::Unverified {
+            return false;
+        }
+        self.status = ProposalStatus::Verified;
+        true
+    }
+
+    #[cfg(test)]
+    fn has_requested_verify(&self) -> bool {
+        self.requested_verify
+    }
```

Comment:
Nit: is this function needed? Is the `requested_verify ` field not visible within the testing in this file?